<a href="https://colab.research.google.com/github/praveen555/IMI-BIGDATA--CHALLENGE/blob/main/UofT_nodes_processing_v1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The main objective of this notebook is to do basic cleaning and pre-processing of categorical columnns. 

In [1]:
import pandas as pd

In [2]:
df_occupation_risk=pd.read_csv('UofT_occupation_risk.csv')

In [3]:
df_occupation_risk

code occupation_risk
0       0             Low
1       1            High
2       2             Low
3       3             Low
4       4             Low
..    ...             ...
836   836             Low
837   837             Low
838   838             Low
839   839            High
840   840            High

[841 rows x 2 columns]

get the distribution counnts

In [4]:
df_occupation_risk['occupation_risk'].value_counts()

Low         629
High        167
Moderate     45
Name: occupation_risk, dtype: int64

In [5]:
df_nodes=pd.read_csv('UofT_nodes.csv',parse_dates=['CUST_ADD_DT','BIRTH_DT'])

In [6]:
df_nodes

BIRTH_DT CUST_ADD_DT  OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA  \
0      1981-09-01  2007-07-05      89.0             1                   1   
1      1994-02-21  2019-05-19      89.0             1                   1   
2      1962-11-16  2011-08-02      89.0             1                   1   
3      1998-06-20  2001-08-15      89.0             1                   1   
4      1942-01-24  2012-10-28      89.0             1                   1   
...           ...         ...       ...           ...                 ...   
999995 1939-01-16  1994-03-19     467.0             0                   1   
999996 1970-12-13  1995-12-22     467.0             0                   1   
999997 1958-09-09  2010-11-03      65.0             0                   1   
999998 1983-09-15  2013-01-30     231.0             0                   1   
999999 1962-01-23  2017-03-07      89.0             0                   0   

        PEP_FL  CASH_SUM_IN  CASH_CNT_IN  CASH_SUM_OUT  CASH_CNT_OUT  \
0          0.0     2577.785          2.0      2851.663           7.0   
1          0.0     3035.502          2.0      4805.997          15.0   
2          0.0     1617.571          1.0      3482.809          11.0   
3          0.0     3587.042          5.0      1940.943           8.0   
4          0.0     1725.524          2.0     13197.169          43.0   
...        ...          ...          ...           ...           ...   
999995     0.0    18945.108         10.0     32210.000          21.0   
999996     0.0     5775.601          7.0      5015.741          15.0   
999997     0.0     2351.255          2.0     15997.000          19.0   
999998     0.0    27312.613          8.0      3399.000           5.0   
999999     0.0     1421.363          1.0      1276.000           1.0   

        WIRES_SUM_IN  WIRES_CNT_IN  WIRES_SUM_OUT  WIRES_CNT_OUT  \
0            70774.0          62.0        24244.0           35.0   
1           282929.0          76.0       196439.0          152.0   
2             7632.0           6.0        33598.0           20.0   
3           149253.0          71.0       105896.0          238.0   
4            92318.0          96.0       187403.0          180.0   
...              ...           ...            ...            ...   
999995     1186467.0          25.0      2550098.0          174.0   
999996      244504.0          53.0       417992.0          337.0   
999997     3024271.0          68.0       948981.0          124.0   
999998      943477.0           9.0       376429.0           12.0   
999999      683797.0           5.0       617628.0           12.0   

       COUNTRY_RISK_INCOME COUNTRY_RISK_RESIDENCY  RISK  \
0                      Low                    Low   low   
1                      Low                    Low   low   
2                      Low                    Low   low   
3                      Low                    Low   low   
4                      Low                    Low   low   
...                    ...                    ...   ...   
999995                 Low                   High  high   
999996                 Low                    Low   low   
999997                 Low                   High  high   
999998                 Low                   High  high   
999999                High                   High  high   

                                  NAME  GENDER  CUSTOMER_ID  
0         Young, Marie Mildren Coleman  Female     935382.0  
1                     Mark Stupar Lecy    Male       2305.0  
2           Dean Glasper Wendel Reeves    Male     472403.0  
3                  Hulsey, Linda Rauth  Female     227994.0  
4           Carolyn Washington Roberts  Female     418265.0  
...                                ...     ...          ...  
999995               Bret* Dove Gainey    Male     420884.0  
999996       Anthony Ray Montano Gomez    Male     429418.0  
999997  Bernadette Schofield Pace Mile  Female     731360.0  
999998         Wasson, Russell Terrell    Male     837988.0  
999999     

In [7]:
df_nodes.isna().sum()

BIRTH_DT                     0
CUST_ADD_DT               4178
OCPTN_NM                  6384
RES_CNTRY_CA                 0
CNTRY_OF_INCOME_CA           0
PEP_FL                       0
CASH_SUM_IN                  0
CASH_CNT_IN                  0
CASH_SUM_OUT                 0
CASH_CNT_OUT                 0
WIRES_SUM_IN                 0
WIRES_CNT_IN                 0
WIRES_SUM_OUT                0
WIRES_CNT_OUT                0
COUNTRY_RISK_INCOME          0
COUNTRY_RISK_RESIDENCY       0
RISK                         0
NAME                       660
GENDER                     628
CUSTOMER_ID                  0
dtype: int64

Rename column

In [8]:
df_occupation_risk.rename(columns={'code':'OCPTN_NM'},inplace=True)

For missing values 
1. Occupation number fill with the highest distribution or even try to even sample with top 5 numbers for better distribution ? But we need to check a significant question is this columns somehow correalated with another column to fill it better ? Current assumption is to occupation code corresponds directly corresponds the risk level label. 
2. Missing gender could be determined by evaluating on male/female names or doing a 50:50 split ( 314 males and 315 females) 
3. Missing names could be discarded if we decide not to use this feature 
4. Have to think about cust_add_data. (if we find a patter from cust_add_date then fill with the most common one).

let us try to understand the NaN ocptn num before filling it.

In [9]:
data=df_nodes[df_nodes['OCPTN_NM'].isna()]

In [10]:
data

BIRTH_DT CUST_ADD_DT  OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA  \
97     1951-10-05  1998-05-15       NaN             1                   1   
124    2010-06-15  2015-08-03       NaN             1                   1   
507    2003-11-10  2014-05-15       NaN             1                   1   
663    2017-04-22  2022-04-26       NaN             1                   1   
710    1939-07-13  1989-04-29       NaN             1                   1   
...           ...         ...       ...           ...                 ...   
999470 1975-04-25  1998-09-22       NaN             0                   1   
999676 1979-01-18  2022-06-06       NaN             0                   0   
999748 1959-10-15  2003-03-28       NaN             0                   1   
999754 1983-05-20  2013-12-11       NaN             0                   1   
999855 1972-01-13  2006-06-07       NaN             0                   1   

        PEP_FL  CASH_SUM_IN  CASH_CNT_IN  CASH_SUM_OUT  CASH_CNT_OUT  \
97         0.0     4480.322          4.0      2367.785           5.0   
124        0.0     2676.499          4.0      8585.831          30.0   
507        0.0     3385.600          5.0     14294.903          47.0   
663        0.0     3404.904          8.0     14019.012          39.0   
710        0.0     2831.035          3.0      6774.756          17.0   
...        ...          ...          ...           ...           ...   
999470     0.0     9820.000          7.0     53401.000          36.0   
999676     0.0    27441.000         11.0     17150.000          12.0   
999748     0.0     7522.000          3.0     69285.000          46.0   
999754     0.0    37888.000         16.0      2763.000           2.0   
999855     0.0    14916.000          5.0     16493.000          10.0   

        WIRES_SUM_IN  WIRES_CNT_IN  WIRES_SUM_OUT  WIRES_CNT_OUT  \
97          402406.0         119.0       366594.0          185.0   
124         398804.0          74.0       301399.0          188.0   
507         219672.0         155.0       269996.0          294.0   
663          20117.0          45.0        38957.0           93.0   
710         221892.0          74.0       349983.0          217.0   
...              ...           ...            ...            ...   
999470     9410086.0          73.0      4513139.0          102.0   
999676     3059300.0          15.0      1436981.0           44.0   
999748    18645082.0          85.0      8807293.0          264.0   
999754     4134466.0          56.0      1880748.0           74.0   
999855      163126.0           6.0       645746.0           24.0   

       COUNTRY_RISK_INCOME COUNTRY_RISK_RESIDENCY    RISK  \
97                     Low                    Low     low   
124                    Low                    Low     low   
507                    Low                    Low     low   
663                    Low                    Low     low   
710                    Low                    Low     low   
...                    ...                    ...     ...   
999470                 Low               Moderate  medium   
999676                 Low               Moderate  medium   
999748                 Low                   High  medium   
999754                 Low                   High  medium   
999855                 Low                    Low  medium   

                                     NAME  GENDER  CUSTOMER_ID  
97              Harris, Donnie Mcpeattehr    Male      72723.0  
124        Mcclure, Richard Evans Fennell    Male     628763.0  
507            Michael Mecham Howse Adams    Male      34995.0  
663     Elizabeth Forman Garrity Marchand  Female     955196.0  
710                  Cole, David Prescott    Male        904.0  
...                                   ...     ...          ...  
999470             Patricia White Vasquez  Female     701136.0  
999676        Harris, Barbara Hildebrandt  Female     220596.0  
999748               Celeste Harris Potts  Female      89038.0  
999754             

get the distribution of the risk values

In [11]:
data['RISK'].value_counts()

low       3861
medium    2199
high       324
Name: RISK, dtype: int64

Can we fill it the value like ??

*low- 3861 so 3861 low values from occupation risk
*medium- 2199 moderate value from occupation risk
*high- 324 high values from occupation risk

In [12]:
data.describe()

OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA       PEP_FL    CASH_SUM_IN  \
count       0.0   6384.000000         6384.000000  6384.000000    6384.000000   
mean        NaN      0.985589            0.986999     0.092732   14694.559952   
std         NaN      0.119187            0.113288     0.290079   19844.917961   
min         NaN      0.000000            0.000000     0.000000       0.410000   
25%         NaN      1.000000            1.000000     0.000000    2611.236000   
50%         NaN      1.000000            1.000000     0.000000    5898.758500   
75%         NaN      1.000000            1.000000     0.000000   18350.585750   
max         NaN      1.000000            1.000000     1.000000  141344.000000   

       CASH_CNT_IN   CASH_SUM_OUT  CASH_CNT_OUT  WIRES_SUM_IN  WIRES_CNT_IN  \
count  6384.000000    6384.000000   6384.000000  6.384000e+03   6384.000000   
mean      8.468358   23008.344852     23.124217  3.323267e+06     84.994518   
std       8.254487   32635.976363     20.641492  5.627575e+06     65.822195   
min       0.000000       0.705000      0.000000  1.700000e+01      0.000000   
25%       3.000000    3536.053250      8.000000  1.207620e+05     34.000000   
50%       6.000000    7988.975500     18.000000  2.923980e+05     71.000000   
75%      11.000000   29116.250000     31.000000  4.527142e+06    120.000000   
max      76.000000  231752.000000    172.000000  3.794526e+07    484.000000   

       WIRES_SUM_OUT  WIRES_CNT_OUT   CUSTOMER_ID  
count   6.384000e+03    6384.000000  6.384000e+03  
mean    1.370194e+06     154.453477  5.009156e+05  
std     2.150274e+06     120.218274  2.891102e+05  
min     6.600000e+01       0.000000  1.290000e+02  
25%     1.081190e+05      61.000000  2.499662e+05  
50%     2.551430e+05     129.000000  4.985170e+05  
75%     1.905200e+06     219.000000  7.520995e+05  
max     1.458973e+07     800.000000  1.000621e+06

check on the distribution of occupation_number before proceeding to fill. 

In [13]:
df_nodes['OCPTN_NM'].value_counts()

25.0     185293
724.0    147064
89.0     135796
513.0    124901
817.0    100147
          ...  
703.0         1
295.0         1
628.0         1
824.0         1
184.0         1
Name: OCPTN_NM, Length: 682, dtype: int64

In [14]:
df_occupation_risk.iloc[[25,724,89,513,817]]

OCPTN_NM occupation_risk
25         25             Low
724       724             Low
89         89            High
513       513            High
817       817             Low

so top 5 occupation labels are low and high. 

he already have 25 for low and 89 for high. Let us get one number for moderate as well. It does now really matter since we have to map to high,medium and low anyways later. 

In [15]:
df_occupation_risk[df_occupation_risk['occupation_risk']=='Moderate']

OCPTN_NM occupation_risk
7           7        Moderate
44         44        Moderate
49         49        Moderate
63         63        Moderate
77         77        Moderate
113       113        Moderate
122       122        Moderate
152       152        Moderate
159       159        Moderate
187       187        Moderate
213       213        Moderate
221       221        Moderate
266       266        Moderate
270       270        Moderate
280       280        Moderate
295       295        Moderate
300       300        Moderate
336       336        Moderate
348       348        Moderate
379       379        Moderate
384       384        Moderate
403       403        Moderate
417       417        Moderate
425       425        Moderate
521       521        Moderate
534       534        Moderate
588       588        Moderate
630       630        Moderate
631       631        Moderate
633       633        Moderate
655       655        Moderate
676       676        Moderate
714       714        Moderate
718       718        Moderate
732       732        Moderate
759       759        Moderate
774       774        Moderate
788       788        Moderate
803       803        Moderate
805       805        Moderate
807       807        Moderate
809       809        Moderate
814       814        Moderate
815       815        Moderate
823       823        Moderate

So mapping numbers are 

25-low,
89- high,
8- medium.

Let us try to get the index values so that we can map it.

In [16]:
low_risk=data[data['RISK']=='low']

In [17]:
low_risk_index=low_risk.index

In [18]:
low_risk_index

Int64Index([    97,    124,    507,    663,    710,    843,   1031,   1063,
              1145,   1252,
            ...
            982163, 982198, 982266, 982279, 983097, 983170, 983273, 983301,
            983592, 985861],
           dtype='int64', length=3861)

In [19]:
medium_risk=data[data['RISK']=='medium']
medium_risk_index=medium_risk.index
high_risk=data[data['RISK']=='high']
high_risk_index=high_risk.index

In [20]:
sum([len(low_risk_index),len(medium_risk),len(high_risk_index)])

6384

above are exact entries which matches with the original Nan counts. 

In [21]:
df_nodes.loc[low_risk_index,'OCPTN_NM']=25

In [22]:
df_nodes.loc[medium_risk_index,'OCPTN_NM']=8

In [23]:
df_nodes.loc[high_risk_index,'OCPTN_NM']=89

In [24]:
df_nodes.isna().sum()

BIRTH_DT                     0
CUST_ADD_DT               4178
OCPTN_NM                     0
RES_CNTRY_CA                 0
CNTRY_OF_INCOME_CA           0
PEP_FL                       0
CASH_SUM_IN                  0
CASH_CNT_IN                  0
CASH_SUM_OUT                 0
CASH_CNT_OUT                 0
WIRES_SUM_IN                 0
WIRES_CNT_IN                 0
WIRES_SUM_OUT                0
WIRES_CNT_OUT                0
COUNTRY_RISK_INCOME          0
COUNTRY_RISK_RESIDENCY       0
RISK                         0
NAME                       660
GENDER                     628
CUSTOMER_ID                  0
dtype: int64

we have now filled occupation num based on the assumption of risk directly being realted to the risk label. 

## Mapping Occupation number for getting categorical values. 

In [25]:
mapping=dict(df_occupation_risk[['OCPTN_NM','occupation_risk']].values)

In [26]:
df_nodes['OCPTN_NM']=df_nodes['OCPTN_NM'].map(mapping)

In [27]:
df_nodes

BIRTH_DT CUST_ADD_DT OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA  \
0      1981-09-01  2007-07-05     High             1                   1   
1      1994-02-21  2019-05-19     High             1                   1   
2      1962-11-16  2011-08-02     High             1                   1   
3      1998-06-20  2001-08-15     High             1                   1   
4      1942-01-24  2012-10-28     High             1                   1   
...           ...         ...      ...           ...                 ...   
999995 1939-01-16  1994-03-19      Low             0                   1   
999996 1970-12-13  1995-12-22      Low             0                   1   
999997 1958-09-09  2010-11-03     High             0                   1   
999998 1983-09-15  2013-01-30     High             0                   1   
999999 1962-01-23  2017-03-07     High             0                   0   

        PEP_FL  CASH_SUM_IN  CASH_CNT_IN  CASH_SUM_OUT  CASH_CNT_OUT  \
0          0.0     2577.785          2.0      2851.663           7.0   
1          0.0     3035.502          2.0      4805.997          15.0   
2          0.0     1617.571          1.0      3482.809          11.0   
3          0.0     3587.042          5.0      1940.943           8.0   
4          0.0     1725.524          2.0     13197.169          43.0   
...        ...          ...          ...           ...           ...   
999995     0.0    18945.108         10.0     32210.000          21.0   
999996     0.0     5775.601          7.0      5015.741          15.0   
999997     0.0     2351.255          2.0     15997.000          19.0   
999998     0.0    27312.613          8.0      3399.000           5.0   
999999     0.0     1421.363          1.0      1276.000           1.0   

        WIRES_SUM_IN  WIRES_CNT_IN  WIRES_SUM_OUT  WIRES_CNT_OUT  \
0            70774.0          62.0        24244.0           35.0   
1           282929.0          76.0       196439.0          152.0   
2             7632.0           6.0        33598.0           20.0   
3           149253.0          71.0       105896.0          238.0   
4            92318.0          96.0       187403.0          180.0   
...              ...           ...            ...            ...   
999995     1186467.0          25.0      2550098.0          174.0   
999996      244504.0          53.0       417992.0          337.0   
999997     3024271.0          68.0       948981.0          124.0   
999998      943477.0           9.0       376429.0           12.0   
999999      683797.0           5.0       617628.0           12.0   

       COUNTRY_RISK_INCOME COUNTRY_RISK_RESIDENCY  RISK  \
0                      Low                    Low   low   
1                      Low                    Low   low   
2                      Low                    Low   low   
3                      Low                    Low   low   
4                      Low                    Low   low   
...                    ...                    ...   ...   
999995                 Low                   High  high   
999996                 Low                    Low   low   
999997                 Low                   High  high   
999998                 Low                   High  high   
999999                High                   High  high   

                                  NAME  GENDER  CUSTOMER_ID  
0         Young, Marie Mildren Coleman  Female     935382.0  
1                     Mark Stupar Lecy    Male       2305.0  
2           Dean Glasper Wendel Reeves    Male     472403.0  
3                  Hulsey, Linda Rauth  Female     227994.0  
4           Carolyn Washington Roberts  Female     418265.0  
...                                ...     ...          ...  
999995               Bret* Dove Gainey    Male     420884.0  
999996       Anthony Ray Montano Gomez    Male     429418.0  
999997  Bernadette Schofield Pace Mile  Female     731360.0  
999998         Wasson, Russell Terrell    Male     837988.0  
999999                D

## Filling Gender Values

In [28]:
df_nodes['GENDER'].value_counts()

Female    548684
Male      450688
Name: GENDER, dtype: int64

In ideal case we would have filled it by trying to recognize the gender from the name. However since there are 628 names and only two types of gender. We will proceed with 50:50 split for the sake of it. 

In [29]:
data=df_nodes[df_nodes['GENDER'].isna()]

In [30]:
nan_gender=data.index

In [31]:
nan_male=nan_gender[:314]
nan_female=nan_gender[314:]

In [32]:
df_nodes.loc[nan_male,'GENDER'] ='Male'

In [33]:
df_nodes.loc[nan_female,'GENDER']='Female'

In [34]:
df_nodes.isna().sum()

BIRTH_DT                     0
CUST_ADD_DT               4178
OCPTN_NM                     0
RES_CNTRY_CA                 0
CNTRY_OF_INCOME_CA           0
PEP_FL                       0
CASH_SUM_IN                  0
CASH_CNT_IN                  0
CASH_SUM_OUT                 0
CASH_CNT_OUT                 0
WIRES_SUM_IN                 0
WIRES_CNT_IN                 0
WIRES_SUM_OUT                0
WIRES_CNT_OUT                0
COUNTRY_RISK_INCOME          0
COUNTRY_RISK_RESIDENCY       0
RISK                         0
NAME                       660
GENDER                       0
CUSTOMER_ID                  0
dtype: int64

## Filling out Cust Add date

Analyse the columns to see if we can find any patterns for the cust_add_Date to fill.

In [35]:
df_nodes['CUST_ADD_DT'].min()

Timestamp('1949-02-10 00:00:00')

In [36]:
df_nodes['CUST_ADD_DT'].max()

Timestamp('2022-07-26 00:00:00')

In [37]:
df_nodes['CUST_ADD_DT'].value_counts()

2008-03-30    140
2006-08-29    139
2003-04-27    138
2009-03-04    134
2005-04-03    134
             ... 
1967-09-30      1
1978-03-24      1
1967-02-11      1
1973-06-21      1
1973-06-13      1
Name: CUST_ADD_DT, Length: 19411, dtype: int64

19411 unique cust_add_date with 2008-03-30 being the most common. let us look at type of customer again which are having missing values. 

In [38]:
data=df_nodes[df_nodes['CUST_ADD_DT'].isna()]

In [39]:
data['RISK'].value_counts()

low       2548
medium    1455
high       175
Name: RISK, dtype: int64

One idea is just select a random date from each of these categories and fill it up. Could be find better correlations to accurately fill up ? Need to discuss this.

Another idea is to find the average average age between birth_date and bank opening date and just add that value. But this is could be more vauge ? 

For example people usually by age of 20 years have a bank accounts. So their DOB+20 years could give us the cust_add_Date. for more accuracy we might need to perform average for month, day and year seperarely also and dig deeper for each category.

In [40]:
data

BIRTH_DT CUST_ADD_DT OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA  \
483    1987-04-15         NaT     High             1                   1   
495    1992-12-07         NaT     High             1                   1   
570    1979-03-12         NaT     High             1                   1   
671    1961-08-26         NaT     High             1                   1   
673    1956-12-19         NaT     High             1                   1   
...           ...         ...      ...           ...                 ...   
998929 1952-10-16         NaT     High             0                   1   
998977 1958-03-24         NaT      Low             0                   1   
999399 1948-08-01         NaT     High             0                   1   
999712 1936-06-17         NaT      Low             0                   1   
999735 1960-07-13         NaT     High             0                   1   

        PEP_FL  CASH_SUM_IN  CASH_CNT_IN  CASH_SUM_OUT  CASH_CNT_OUT  \
483        0.0     1759.401          3.0      1906.922           4.0   
495        0.0     6520.571          8.0     11431.672          35.0   
570        0.0      764.322          1.0      4986.825          15.0   
671        0.0     4888.680          7.0       905.752           4.0   
673        0.0     2099.690          4.0      2532.592           5.0   
...        ...          ...          ...           ...           ...   
998929     0.0    21257.000          7.0      1850.000           1.0   
998977     0.0    18888.000         11.0    137969.000          72.0   
999399     0.0     2446.000          2.0    101240.000          65.0   
999712     0.0    16994.000         11.0     71988.000          46.0   
999735     0.0    32562.000         10.0     24582.000           7.0   

        WIRES_SUM_IN  WIRES_CNT_IN  WIRES_SUM_OUT  WIRES_CNT_OUT  \
483         332432.0         141.0       302831.0          289.0   
495         277602.0         155.0       344764.0          316.0   
570         309263.0          90.0       150943.0           71.0   
671          21855.0          17.0       134618.0           90.0   
673         197005.0          26.0       183271.0           46.0   
...              ...           ...            ...            ...   
998929     3551256.0          18.0       830042.0           23.0   
998977    14888830.0         162.0      4303582.0          309.0   
999399    32659741.0         211.0      9923733.0          193.0   
999712    21777373.0          41.0      6629770.0          103.0   
999735     3928840.0          63.0      2544611.0          143.0   

       COUNTRY_RISK_INCOME COUNTRY_RISK_RESIDENCY    RISK  \
483                    Low                    Low     low   
495                    Low                    Low     low   
570                    Low                    Low     low   
671                    Low                    Low     low   
673                    Low                    Low     low   
...                    ...                    ...     ...   
998929                 Low                    Low  medium   
998977                 Low                    Low  medium   
999399                 Low                   High  medium   
999712                 Low                   High  medium   
999735                 Low                   High  medium   

                                  NAME  GENDER  CUSTOMER_ID  
483     Maria Riahcrdson Stanley Perez  Female     421063.0  
495               Laura Ransler Sexton  Female     282698.0  
570              Jon Lden Bunch Bryant    Male     767922.0  
671     Lisa Senecal Blacklock Steffes  Female     488603.0  
673      Francisco Perry Grant Mattila    Male     349364.0  
...                                ...     ...          ...  
998929  Fender, JJimy Dodge Warrington    Male     654549.0  
998977         Hnde1on, Jeanne Munguia  Female     214776.0  
999399          Rebecca Taylor Graeser  Female     712953.0  
999712           Joyce Mccormack Reese  Female     276565.0  

In [41]:
data2=df_nodes[df_nodes['CUST_ADD_DT'].notna()]

In [42]:
data2['year_dob']=data2['BIRTH_DT'].dt.year
data2['month_dob']=data2['BIRTH_DT'].dt.month
data2['day_dob']=data2['BIRTH_DT'].dt.day
data2['cust_add_year']=data2['CUST_ADD_DT'].dt.year
data2['cust_add_month']=data2['CUST_ADD_DT'].dt.month
data2['cust_add_day']=data2['CUST_ADD_DT'].dt.day

<ipython-input-42-fbea2e8fa120>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data2['year_dob']=data2['BIRTH_DT'].dt.year
<ipython-input-42-fbea2e8fa120>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data2['month_dob']=data2['BIRTH_DT'].dt.month
<ipython-input-42-fbea2e8fa120>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/

In [43]:
data2['year_diff']=data2['cust_add_year']-data2['year_dob']


<ipython-input-43-307f94cb6d68>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data2['year_diff']=data2['cust_add_year']-data2['year_dob']


In [44]:
data2['month_diff']=abs(data2['cust_add_month']-data2['month_dob'])
data2['day_diff']=abs(data2['cust_add_day']-data2['day_dob'])

<ipython-input-44-78a7a846a857>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data2['month_diff']=abs(data2['cust_add_month']-data2['month_dob'])
<ipython-input-44-78a7a846a857>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data2['day_diff']=abs(data2['cust_add_day']-data2['day_dob'])


try to find average by risk category because feature should ultimately correlate to the column. 

In [45]:
data_high=data2[data2['RISK']=='high']

In [46]:
data_high

BIRTH_DT CUST_ADD_DT OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA  \
123292 1977-04-21  1995-02-11     High             1                   1   
123293 1995-04-07  2003-12-15     High             1                   1   
123294 1956-08-01  1999-12-17     High             1                   1   
123295 1959-02-14  2006-08-08     High             1                   1   
123296 1981-05-29  2016-05-02     High             1                   1   
...           ...         ...      ...           ...                 ...   
999994 1970-08-21  2014-09-09      Low             0                   1   
999995 1939-01-16  1994-03-19      Low             0                   1   
999997 1958-09-09  2010-11-03     High             0                   1   
999998 1983-09-15  2013-01-30     High             0                   1   
999999 1962-01-23  2017-03-07     High             0                   0   

        PEP_FL  CASH_SUM_IN  CASH_CNT_IN  CASH_SUM_OUT  CASH_CNT_OUT  ...  \
123292     0.0     9622.865          1.0        4898.0           5.0  ...   
123293     0.0    19651.804         11.0       28033.0          14.0  ...   
123294     0.0    25831.652         11.0        8488.0           2.0  ...   
123295     0.0    36870.662         16.0       17456.0          15.0  ...   
123296     0.0    67572.085         25.0       34554.0          22.0  ...   
...        ...          ...          ...           ...           ...  ...   
999994     0.0    21161.905          8.0       30217.0          24.0  ...   
999995     0.0    18945.108         10.0       32210.0          21.0  ...   
999997     0.0     2351.255          2.0       15997.0          19.0  ...   
999998     0.0    27312.613          8.0        3399.0           5.0  ...   
999999     0.0     1421.363          1.0        1276.0           1.0  ...   

        CUSTOMER_ID  year_dob  month_dob  day_dob cust_add_year  \
123292     132001.0      1977          4       21          1995   
123293     280713.0      1995          4        7          2003   
123294     214538.0      1956          8        1          1999   
123295     345313.0      1959          2       14          2006   
123296      52253.0      1981          5       29          2016   
...             ...       ...        ...      ...           ...   
999994     377454.0      1970          8       21          2014   
999995     420884.0      1939          1       16          1994   
999997     731360.0      1958          9        9          2010   
999998     837988.0      1983          9       15          2013   
999999     166357.0      1962          1       23          2017   

       cust_add_month cust_add_day year_diff month_diff  day_diff  
123292              2           11        18          2        10  
123293             12           15         8          8         8  
123294             12           17        43          4        16  
123295              8            8        47          6         6  
123296              5            2        35          0        27  
...               ...          ...       ...        ...       ...  
999994              9            9        44          1        12  
999995              3           19        55          2         3  
999997             11            3        52          2         6  
999998              1           30        30          8        15  
999999              3            7        55          2        16  

[49825 rows x 29 columns]

for high risk

In [47]:
print('avg year diff',round(data_high['year_diff'].mean()))
print('avg month diff',round(data_high['month_diff'].mean()))
print('avg day diff',round(data_high['day_diff'].mean()))

avg year diff 35
avg month diff 4
avg day diff 10


for low risk

In [48]:
data_low=data2[data2['RISK']=='low']

In [49]:
print('avg year diff',round(data_low['year_diff'].mean()))
print('avg month diff',round(data_low['month_diff'].mean()))
print('avg day diff',round(data_low['day_diff'].mean()))

avg year diff 31
avg month diff 4
avg day diff 10


for medium risk

In [50]:
data_medium=data2[data2['RISK']=='medium']

In [51]:
data_medium

BIRTH_DT CUST_ADD_DT OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA  \
60907  1977-02-09  1984-06-09     High             1                   1   
60908  1948-02-03  1998-12-23     High             1                   1   
60909  1952-09-29  1985-12-17     High             1                   1   
60910  1978-09-27  2001-09-09     High             1                   1   
60911  1958-08-01  2021-04-26     High             1                   1   
...           ...         ...      ...           ...                 ...   
999871 1977-11-16  2021-02-18      Low             0                   1   
999872 1976-10-07  1996-12-12      Low             0                   1   
999873 1958-12-21  2008-11-27      Low             0                   1   
999874 1996-09-24  2022-03-25      Low             0                   1   
999875 1940-02-20  1997-08-28      Low             0                   0   

        PEP_FL  CASH_SUM_IN  CASH_CNT_IN  CASH_SUM_OUT  CASH_CNT_OUT  ...  \
60907      0.0      40341.0         20.0       49774.0          34.0  ...   
60908      1.0      16132.0          8.0       15399.0           6.0  ...   
60909      1.0      48487.0         21.0       56489.0          41.0  ...   
60910      0.0      32847.0         13.0       71896.0          41.0  ...   
60911      0.0      20107.0          6.0       33453.0          20.0  ...   
...        ...          ...          ...           ...           ...  ...   
999871     0.0      40869.0         18.0       47952.0          33.0  ...   
999872     0.0      21775.0         12.0       38097.0          27.0  ...   
999873     1.0       7659.0          3.0       21357.0          15.0  ...   
999874     0.0      18043.0          6.0       87439.0          60.0  ...   
999875     0.0      13099.0          5.0       75244.0          34.0  ...   

        CUSTOMER_ID  year_dob  month_dob  day_dob cust_add_year  \
60907      801242.0      1977          2        9          1984   
60908       85429.0      1948          2        3          1998   
60909      618775.0      1952          9       29          1985   
60910      173546.0      1978          9       27          2001   
60911      182874.0      1958          8        1          2021   
...             ...       ...        ...      ...           ...   
999871     860132.0      1977         11       16          2021   
999872     836967.0      1976         10        7          1996   
999873    1000248.0      1958         12       21          2008   
999874    1000630.0      1996          9       24          2022   
999875     477516.0      1940          2       20          1997   

       cust_add_month cust_add_day year_diff month_diff  day_diff  
60907               6            9         7          4         0  
60908              12           23        50         10        20  
60909              12           17        33          3        12  
60910               9            9        23          0        18  
60911               4           26        63          4        25  
...               ...          ...       ...        ...       ...  
999871              2           18        44          9         2  
999872             12           12        20          2         5  
999873             11           27        50          1         6  
999874              3           25        26          6         1  
999875              8           28        57          6         8  

[348545 rows x 29 columns]

In [52]:
print('avg year diff',round(data_medium['year_diff'].mean()))
print('avg month diff',round(data_medium['month_diff'].mean()))
print('avg day diff',round(data_medium['day_diff'].mean()))

avg year diff 35
avg month diff 4
avg day diff 10


In [53]:
data2['cust_add_year'].value_counts()

2006    37791
2005    37600
2007    37409
2004    37085
2008    36927
        ...  
1955        2
1954        2
1950        1
1951        1
1949        1
Name: cust_add_year, Length: 72, dtype: int64

### Why is the average of all cust_add_dt around 4 for months and 10 for days ? 



*   Another pattern which we can observe is the average years in both medium and high risk is 35 years from their DOB




In [54]:
print('average of all cust--year',data2['cust_add_year'].mean())
print('average of all cust--month',data2['cust_add_month'].mean())
print('average of all cust--day',data2['cust_add_day'].mean())
print('average of all dob--year',data2['year_dob'].mean())
print('average of all dob--month',data2['month_dob'].mean())
print('average of all dob--day',data2['day_dob'].mean())





average of all cust--year 2004.1107697962086
average of all cust--month 6.519600892528986
average of all cust--day 15.716023546376762
average of all dob--year 1971.6695684570134
average of all dob--month 6.525401125904027
average of all dob--day 15.740675542416215


As it check out average difference in DOB is same. The month is 6 and day is 15. 

The average age between DOB and customer_add_date is 2004-1971 approx 33 years. 

Can we proceed by filling the above one based on these stats. 

Discuss all these analysis with team. 

In [55]:
from datetime import datetime, timedelta
from dateutil.relativedelta import *

In [56]:
date= datetime(2000,1,1)
date = date + relativedelta(years=+33,months=+6,days=+10)
print(date.date())

2033-07-11


In [57]:
def filldates_high_medium(dob_date):
  cust_add_dt=dob_date+relativedelta(years=+35,months=+4,days=+10)
  return cust_add_dt.date()

In [58]:
def filldates_low(dob_date):
  cust_add_dt=dob_date+relativedelta(years=+31,months=+4,days=+10)
  return cust_add_dt.date()

In [59]:
high_medium=data[(data['RISK']=='high') | (data['RISK']=='medium')]

In [60]:
idx_high_medium=high_medium.index

In [61]:
low=data[(data['RISK']=='low')]
idx_low=low.index

In [62]:
df_nodes.loc[idx_high_medium, 'CUST_ADD_DT'] = df_nodes.loc[idx_high_medium, 'BIRTH_DT'].apply(filldates_high_medium)

In [63]:
df_nodes.loc[idx_low, 'CUST_ADD_DT'] = df_nodes.loc[idx_low, 'BIRTH_DT'].apply(filldates_low)

In [64]:
df_nodes.isna().sum()

BIRTH_DT                    0
CUST_ADD_DT                 0
OCPTN_NM                    0
RES_CNTRY_CA                0
CNTRY_OF_INCOME_CA          0
PEP_FL                      0
CASH_SUM_IN                 0
CASH_CNT_IN                 0
CASH_SUM_OUT                0
CASH_CNT_OUT                0
WIRES_SUM_IN                0
WIRES_CNT_IN                0
WIRES_SUM_OUT               0
WIRES_CNT_OUT               0
COUNTRY_RISK_INCOME         0
COUNTRY_RISK_RESIDENCY      0
RISK                        0
NAME                      660
GENDER                      0
CUSTOMER_ID                 0
dtype: int64

## Questions on names ? 

it is best to discard this we are not going to anyways use this column.

## Mapping categorical columns 

Since this more based on ranking system such as high,medium and low. Ordinal encoding could be done instead of one hot encoding as there exists a relationship.

In [65]:
df_nodes.columns

Index(['BIRTH_DT', 'CUST_ADD_DT', 'OCPTN_NM', 'RES_CNTRY_CA',
       'CNTRY_OF_INCOME_CA', 'PEP_FL', 'CASH_SUM_IN', 'CASH_CNT_IN',
       'CASH_SUM_OUT', 'CASH_CNT_OUT', 'WIRES_SUM_IN', 'WIRES_CNT_IN',
       'WIRES_SUM_OUT', 'WIRES_CNT_OUT', 'COUNTRY_RISK_INCOME',
       'COUNTRY_RISK_RESIDENCY', 'RISK', 'NAME', 'GENDER', 'CUSTOMER_ID'],
      dtype='object')

In [66]:
df_nodes['COUNTRY_RISK_RESIDENCY'].value_counts()

Low         990109
High          9200
Moderate       691
Name: COUNTRY_RISK_RESIDENCY, dtype: int64

In [67]:
mapper1={'low':0,'medium':1,'high':2}
mapper2={'Low':0,'Moderate':1,'High':2}


In [68]:
df_nodes['RISK']=df_nodes['RISK'].map(mapper1)

In [69]:
df_nodes['COUNTRY_RISK_RESIDENCY']=df_nodes['COUNTRY_RISK_RESIDENCY'].map(mapper2)
df_nodes['OCPTN_NM']=df_nodes['OCPTN_NM'].map(mapper2)

In [70]:
df_nodes['COUNTRY_RISK_INCOME']=df_nodes['COUNTRY_RISK_INCOME'].map(mapper2)

## Map Gender 

0- Male 
1- Female

In [71]:
mapper3={'Male':0,"Female":1}

In [72]:
df_nodes['GENDER']=df_nodes['GENDER'].map(mapper3)

In [73]:
df_nodes

BIRTH_DT CUST_ADD_DT  OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA  \
0      1981-09-01  2007-07-05         2             1                   1   
1      1994-02-21  2019-05-19         2             1                   1   
2      1962-11-16  2011-08-02         2             1                   1   
3      1998-06-20  2001-08-15         2             1                   1   
4      1942-01-24  2012-10-28         2             1                   1   
...           ...         ...       ...           ...                 ...   
999995 1939-01-16  1994-03-19         0             0                   1   
999996 1970-12-13  1995-12-22         0             0                   1   
999997 1958-09-09  2010-11-03         2             0                   1   
999998 1983-09-15  2013-01-30         2             0                   1   
999999 1962-01-23  2017-03-07         2             0                   0   

        PEP_FL  CASH_SUM_IN  CASH_CNT_IN  CASH_SUM_OUT  CASH_CNT_OUT  \
0          0.0     2577.785          2.0      2851.663           7.0   
1          0.0     3035.502          2.0      4805.997          15.0   
2          0.0     1617.571          1.0      3482.809          11.0   
3          0.0     3587.042          5.0      1940.943           8.0   
4          0.0     1725.524          2.0     13197.169          43.0   
...        ...          ...          ...           ...           ...   
999995     0.0    18945.108         10.0     32210.000          21.0   
999996     0.0     5775.601          7.0      5015.741          15.0   
999997     0.0     2351.255          2.0     15997.000          19.0   
999998     0.0    27312.613          8.0      3399.000           5.0   
999999     0.0     1421.363          1.0      1276.000           1.0   

        WIRES_SUM_IN  WIRES_CNT_IN  WIRES_SUM_OUT  WIRES_CNT_OUT  \
0            70774.0          62.0        24244.0           35.0   
1           282929.0          76.0       196439.0          152.0   
2             7632.0           6.0        33598.0           20.0   
3           149253.0          71.0       105896.0          238.0   
4            92318.0          96.0       187403.0          180.0   
...              ...           ...            ...            ...   
999995     1186467.0          25.0      2550098.0          174.0   
999996      244504.0          53.0       417992.0          337.0   
999997     3024271.0          68.0       948981.0          124.0   
999998      943477.0           9.0       376429.0           12.0   
999999      683797.0           5.0       617628.0           12.0   

        COUNTRY_RISK_INCOME  COUNTRY_RISK_RESIDENCY  RISK  \
0                         0                       0     0   
1                         0                       0     0   
2                         0                       0     0   
3                         0                       0     0   
4                         0                       0     0   
...                     ...                     ...   ...   
999995                    0                       2     2   
999996                    0                       0     0   
999997                    0                       2     2   
999998                    0                       2     2   
999999                    2                       2     2   

                                  NAME  GENDER  CUSTOMER_ID  
0         Young, Marie Mildren Coleman       1     935382.0  
1                     Mark Stupar Lecy       0       2305.0  
2           Dean Glasper Wendel Reeves       0     472403.0  
3                  Hulsey, Linda Rauth       1     227994.0  
4           Carolyn Washington Roberts       1     418265.0  
...                                ...     ...          ...  
999995               Bret* Dove Gainey       0     420884.0  
999996       Anthony Ray Montano Gomez       0     429418.0  
999997  Bernadette Schofield Pace Mile       1     731360.0  
999998         Wasson, Russell Terrell       0   

In [74]:
df_nodes.isna().sum()

BIRTH_DT                    0
CUST_ADD_DT                 0
OCPTN_NM                    0
RES_CNTRY_CA                0
CNTRY_OF_INCOME_CA          0
PEP_FL                      0
CASH_SUM_IN                 0
CASH_CNT_IN                 0
CASH_SUM_OUT                0
CASH_CNT_OUT                0
WIRES_SUM_IN                0
WIRES_CNT_IN                0
WIRES_SUM_OUT               0
WIRES_CNT_OUT               0
COUNTRY_RISK_INCOME         0
COUNTRY_RISK_RESIDENCY      0
RISK                        0
NAME                      660
GENDER                      0
CUSTOMER_ID                 0
dtype: int64

In [75]:
df_nodes

BIRTH_DT CUST_ADD_DT  OCPTN_NM  RES_CNTRY_CA  CNTRY_OF_INCOME_CA  \
0      1981-09-01  2007-07-05         2             1                   1   
1      1994-02-21  2019-05-19         2             1                   1   
2      1962-11-16  2011-08-02         2             1                   1   
3      1998-06-20  2001-08-15         2             1                   1   
4      1942-01-24  2012-10-28         2             1                   1   
...           ...         ...       ...           ...                 ...   
999995 1939-01-16  1994-03-19         0             0                   1   
999996 1970-12-13  1995-12-22         0             0                   1   
999997 1958-09-09  2010-11-03         2             0                   1   
999998 1983-09-15  2013-01-30         2             0                   1   
999999 1962-01-23  2017-03-07         2             0                   0   

        PEP_FL  CASH_SUM_IN  CASH_CNT_IN  CASH_SUM_OUT  CASH_CNT_OUT  \
0          0.0     2577.785          2.0      2851.663           7.0   
1          0.0     3035.502          2.0      4805.997          15.0   
2          0.0     1617.571          1.0      3482.809          11.0   
3          0.0     3587.042          5.0      1940.943           8.0   
4          0.0     1725.524          2.0     13197.169          43.0   
...        ...          ...          ...           ...           ...   
999995     0.0    18945.108         10.0     32210.000          21.0   
999996     0.0     5775.601          7.0      5015.741          15.0   
999997     0.0     2351.255          2.0     15997.000          19.0   
999998     0.0    27312.613          8.0      3399.000           5.0   
999999     0.0     1421.363          1.0      1276.000           1.0   

        WIRES_SUM_IN  WIRES_CNT_IN  WIRES_SUM_OUT  WIRES_CNT_OUT  \
0            70774.0          62.0        24244.0           35.0   
1           282929.0          76.0       196439.0          152.0   
2             7632.0           6.0        33598.0           20.0   
3           149253.0          71.0       105896.0          238.0   
4            92318.0          96.0       187403.0          180.0   
...              ...           ...            ...            ...   
999995     1186467.0          25.0      2550098.0          174.0   
999996      244504.0          53.0       417992.0          337.0   
999997     3024271.0          68.0       948981.0          124.0   
999998      943477.0           9.0       376429.0           12.0   
999999      683797.0           5.0       617628.0           12.0   

        COUNTRY_RISK_INCOME  COUNTRY_RISK_RESIDENCY  RISK  \
0                         0                       0     0   
1                         0                       0     0   
2                         0                       0     0   
3                         0                       0     0   
4                         0                       0     0   
...                     ...                     ...   ...   
999995                    0                       2     2   
999996                    0                       0     0   
999997                    0                       2     2   
999998                    0                       2     2   
999999                    2                       2     2   

                                  NAME  GENDER  CUSTOMER_ID  
0         Young, Marie Mildren Coleman       1     935382.0  
1                     Mark Stupar Lecy       0       2305.0  
2           Dean Glasper Wendel Reeves       0     472403.0  
3                  Hulsey, Linda Rauth       1     227994.0  
4           Carolyn Washington Roberts       1     418265.0  
...                                ...     ...          ...  
999995               Bret* Dove Gainey       0     420884.0  
999996       Anthony Ray Montano Gomez       0     429418.0  
999997  Bernadette Schofield Pace Mile       1     731360.0  
999998         Wasson, Russell Terrell       0   

In [76]:
from google.colab import files
df_nodes.to_csv('clean_uoft_nodes_v1.csv')
files.download('clean_uoft_nodes_v1.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>